### ISSUE TITLE
No blocked-page

### ISSUE THREAD

**Issue by molikuner** [pi-hole/pi-hole#1598]
- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*
- [x] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the codebase?:**

_{2}_

---
**[BUG | ISSUE] Expected Behaviour:**

when visiting a blocked page a page is shown that says that this page is blocked
the admin page is working

**[BUG | ISSUE] Actual Behaviour:**

when visiting a blocked page (`adfoc.us/..........`) there is a page with nothing
the admin page is working

**[BUG | ISSUE] Steps to reproduce:**

- install Pi-hole on raspbian jessi
- let the default settings
- request `http://adfoc.us/..........`
- now you have a white page

**Comment by technicalpyro** [comment]
i have run into this myself can you run `pihole -r` and there will be a section after you select repair referring to blocking page sometimes it seems to not install correctly

**Comment by molikuner** [comment]
debug token: "3g8mmiogu"

to generate this debug token i needed to disable my firewall!
which port does it use?

and no the firewall isn't the problem. it is also like described when the firewall is disabled

I was running "sudo pihole -d" and selected repair
but that didn't helped

**Comment by dschaper** [comment]
Debug uploads to port 9999. Do you have all traffic egressing from your network firewalled?

**Comment by dschaper** [comment]
The debug log doesn't show any activity from the `dnsmasq` server, which leads us to the conclusion that your clients aren't actually using the Pi-hole for blocking?

**Comment by molikuner** [comment]
I requested the pi-hole IP, which means no dns
But in earlier versions there was an blocked-page also there

I requested a new page: "b3258xo6xn"
there is also no blocked-page (white page)

**Comment by dschaper** [comment]
Try opening the Admin page, and using the tools to Tail Logs, this will show in real-time all queries that are being handled by the Pi-hole, if you see no activity when you use a client to browse a website, then the client is not using the Pi-hole for name resolution and you will not see any block pages.

**Comment by molikuner** [comment]
I use the Pi-hole (configured in network settings and there is sth in the logs)

the page I requested was an ad-link: adfoc.us............

now I requested only http://adfoc.us
there is a blocked-page

**Comment by molikuner** [comment]
I was experimenting a little bit with the url

I found out that the Pi-hole fails to show an blocked-page page if there is a second valid url in the url
for example: http://adfoc.us/google.com

when requesting: http://adfoc.us/google. I can see a blocked-page page

**Comment by dschaper** [comment]
`http://adfoc.us/google.com` is just `http://adfoc.us` with a directory of `google.com`, it's not a combination of two domains, or I'm not following what the question is.

**Comment by molikuner** [comment]
yes that is true but `adfoc.us` is a page on an advertisement list
means I would get a bocked-page page

but when I request `http://adfoc.us/google.com` I get a white page and no blocked-page page

**Comment by dschaper** [comment]
And what is the HTML content of the white page. Can you do a `curl -I http://adfoc.us/google.com` and `curl http://adfoc.us/google.com` and show the result you are seeing?

**Comment by molikuner** [comment]
$ curl -I http://adfoc.us/google.com
HTTP/1.1 200 OK
X-Pi-hole: A black hole for Internet advertisements.
Content-type: text/html; charset=UTF-8
Date: Wed, 12 Jul 2017 20:59:53 GMT
Server: lighttpd/1.4.35

$ curl http://adfoc.us/google.com
<html>
<head>
<script>window.close();</script></head>
<body>
<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7">
</body>
</html>


![screenshot from 2017-07-12 22-57-10](https://user-images.githubusercontent.com/25723781/28139732-9dc54106-6755-11e7-8a9f-f7735b1a7254.png)

**Comment by jacobsalmela** [comment]
I can replicate the issue on my Pi-hole, just not sure what's happening here.

No block page if navigating to`http://adfoc.us/google.com`

```
$ curl -I http://adfoc.us/google.com
HTTP/1.1 200 OK
X-Pi-hole: A black hole for Internet advertisements.
Content-type: text/html; charset=UTF-8
Date: Wed, 12 Jul 2017 21:00:06 GMT
Server: lighttpd/1.4.35
```
but a block page if you visit `http://adfoc.us/google.`
```
$ curl -I http://adfoc.us/google.
HTTP/1.1 200 OK
X-Pi-hole: A black hole for Internet advertisements.
Content-type: text/html; charset=UTF-8
Date: Wed, 12 Jul 2017 20:59:48 GMT
Server: lighttpd/1.4.35
```

**Comment by dschaper** [comment]
What does the `access.log` and `error.log` entries show when you visit those URLs?

**Comment by molikuner** [comment]
where are these logs?

**Comment by dschaper** [comment]
`/var/log/lighttpd`

**Comment by molikuner** [comment]
access.log
`1499893648|adfoc.us|GET /google.com HTTP/1.1|200|179`

error.log
nothing

**Comment by technicalpyro** [comment]
Access log 
```
1499893836|adfoc.us|GET /google. HTTP/1.1|200|4358
1499893836|pi.hole|GET /admin/scripts/pi-hole/js/queryads.js HTTP/1.1|304|0
1499893840|adfoc.us|GET /admin/scripts/pi-hole/php/queryads.php?domain=adfoc.us&exact HTTP/1.1|200|3475
1499893843|adfoc.us|GET /google.com HTTP/1.1|200|179

```
Error log
Nothing

not sure what @molikuner log will show though

**Comment by dschaper** [comment]
Okay, the A records are pointing to the Pi-hole, but the AAAA records are pointing to the real address.

Try `dig adfoc.us A` and `dig adfoc.us AAAA` and then `head -n 10 /etc/pihole/gravity.list`

**Comment by molikuner** [comment]
$ sudo dig adfoc.us A

; <<>> DiG 9.9.5-9+deb8u12-Raspbian <<>> adfoc.us A
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 14655
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;adfoc.us.			IN	A

;; ANSWER SECTION:
adfoc.us.		300	IN	A	192.168.2.35

;; Query time: 3 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Wed Jul 12 23:27:21 CEST 2017
;; MSG SIZE  rcvd: 53


# molikuner @ PC in ~ [23:27:21] 
$ sudo dig adfoc.us AAAA

; <<>> DiG 9.9.5-9+deb8u12-Raspbian <<>> adfoc.us AAAA
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 65100
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;adfoc.us.			IN	AAAA

;; ANSWER SECTION:
adfoc.us.		63	IN	AAAA	2400:cb00:2048:1::6819:af0f
adfoc.us.		63	IN	AAAA	2400:cb00:2048:1::6819:ae0f

;; Query time: 3 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Wed Jul 12 23:27:38 CEST 2017
;; MSG SIZE  rcvd: 93


# molikuner @ PC in ~ [23:27:38] 
$ head -n 10 /etc/pihole/gravity.list

192.168.2.35 0.0.0.0
192.168.2.35 0000mps.webpreview.dsl.net
192.168.2.35 0001.2waky.com
192.168.2.35 000143owaweboutlookappweb.myfreesites.net
192.168.2.35 000dom.revenuedirect.com
192.168.2.35 000free.us
192.168.2.35 000hhjq.rxportalhosting.com
192.168.2.35 000info.com
192.168.2.35 001wen.com

**Comment by dschaper** [comment]
I'll dig deeper and see what I can resolve, but I'm just not finding where the bug/issue could be, possibly in the `lighttpd` rewrite section?

**Comment by dschaper** [comment]
```
dschaper@nanopineo:~⟫ curl http://adfoc.us/google.com
<html>
<head>
<script>window.close();</script></head>
<body>
<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7">
</body>
</html>
```
```
dschaper@nanopineo:~⟫ curl -I http://adfoc.us/google.com
HTTP/1.1 200 OK
X-Pi-hole: A black hole for Internet advertisements.
Content-type: text/html; charset=UTF-8
Date: Thu, 13 Jul 2017 00:00:12 GMT
Server: lighttpd/1.4.35
```

**Comment by dschaper** [comment]
```
cat /var/www/html/pihole/index.php
...
// Handle incoming URI types
if (!$showPage)
{
?>
<html>
<head>
<script>window.close();</script></head>
<body>
<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7">
</body>
</html>
<?php
        die();
}
...
```

So it's internal to the `index.php` file

**Comment by dschaper** [comment]
And
```
// Define which URL extensions get rendered as "Website Blocked"
$webExt = array('asp', 'htm', 'html', 'php', 'rss', 'xml');

if(in_array($uriExt, $webExt) || empty($uriExt))
{
        // Requested resource has an extension listed in $webExt
        // or no extension (index access to some folder incl. the root dir)
        $showPage = true;
}
else
{
        // Something else
        $showPage = false;
}
```

So `.com` is not a recognized extention for a page and it's tripping the check.

**Comment by dschaper** [comment]
@jacobsalmela @WaLLy3K I'd label this as a won'tfix? I can't think of any site that would server a page with an extention that wasn't one of the standard language extentions, and I don't know of a browser that, if fed a `index.com` page would know how to render it anyways?

**Comment by WaLLy3K** [comment]
Currently, the Block Page appears to be working as intended - my understanding of the current setup is that it's seeing a full stop as a blank file extension, and it's being blocked.

I can't reproduce it using my [non-official block page](https://github.com/WaLLy3K/Pi-hole-Block-Page), and therefore I'm confident that a [future official update will resolve this](https://github.com/pi-hole/pi-hole/pull/1416/files#diff-622eb992a761acc873605fcdfbaf1e3bR51). :smile:

**Comment by dschaper** [comment]
That was my understanding `domain.com/google.asp` would block as would `domain.com/google.` but if you try with a nonstandard extension `domain.com/google.com` the `.com` extension would be considered to be bogon and would instead return a page with the base64 encoded blank gif file.

**Comment by molikuner** [comment]
@dschaper 
There are sites out there that using this extension
`adfoc.us/serve/sitelinks/?id=271228&url=google.com`
And because `id` and `url` are only arguments you would see the content of sitelinks (this is a page with advertisements and a button on which you can click after 5 seconds to be redirected to the url argument)

**Comment by molikuner** [comment]
I logged the variables in the index.php:
```
                        uri                                             -     uriExt
/serve/sitelinks/\?id=ANID\&url=google.com     -     com
```
am I right that the `pathinfo` function from php makes some mistakes?

**Comment by WaLLy3K** [comment]
> adfoc.us/serve/sitelinks/?id=ANID&url=google.com

This URL _should_ be handled as a file due to the `?` query string.

**Comment by molikuner** [comment]
the variable `showPage` is `false`
which means that is is handled like a file but the variable `uriExt` is `com` and not empty

when this page is handled like a file then there is sth wrong because it isn't a file it is a valid webpage

**Comment by molikuner** [comment]
I changed the line:
`$uriExt = pathinfo($uri, PATHINFO_EXTENSION)`
to:
`$uriExt = pathinfo(strtok($uri,'\?'), PATHINFO_EXTENSION)`

this splits the URL into two parts:
- first (used) without the query 
- second (not used) the query string

now it is working right in the example

**Comment by Th3M3** [comment]
A few days ago, I also run into this issue.

On a website, i found a url that directs me to [api.smartredirect.de/api_v2/ClickGate.php?...](http://api.smartredirect.de/api_v2/ClickGate.php?p=Z5CZZT9icW&amp;k=c1a81f7d9fafdd73aa42ab9c25e194ef&amp;url=https%3A%2F%2Fwww.sicherheitstest.bsi.de%2F&amp;s=buDL-wa-extDL)
The page was opened in a new tab and closed immediately afterwards. 
It has no content, but only a `window.close()`-script in the html's head of this page.

Is this issue fixed? I still have it in current vDev branch.

**Comment by WaLLy3K** [comment]
[Sounds like expected behaviour](https://github.com/pi-hole/pi-hole/blob/master/advanced/index.php#L53) to me.

**Comment by Th3M3** [comment]
@molikuner's Patch is working.

I have overlooked that his pull request was merged into `pi-hole:release/3.2` branch and not into `development`.

I hope v3.2 will be released soon.

**Comment by graysky2** [comment]
@Th3M3 - Is it just the one patch you believe fixes this issue?  I applied it to v3.1.4 but I do not see the expected "website blocked" page @molikuner posted a few above.  Instead, I am redirected to the pi-hole admin interface.

**Comment by graysky2** [comment]
Using the Arch Linux package for pi-hole causes a similar behavior for me: blocked sites directly browsed to in the browser do not show the pi-holed page, but simply re-direct to the pi-hole admin interface.

If I use the official pihole script in an ubuntu lxc, it works as it should.  I am very interested in hearing opinions as to why.

Using my Arch package (with unexpected behavior):
```
% curl -v http://adfoc.us/google.com
*   Trying 192.168.1.250...
* TCP_NODELAY set
* Connected to adfoc.us (192.168.1.250) port 80 (#0)
> GET /google.com HTTP/1.1
> Host: adfoc.us
> User-Agent: curl/7.56.1
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Server: nginx/1.13.5
< Date: Sun, 29 Oct 2017 14:50:41 GMT
< Content-Type: text/html; charset=UTF-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< X-Powered-By: PHP/7.1.11
< 
<html>
<head>
<script>window.close();</script></head>
<body>
<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7">
</body>
</html>
* Connection #0 to host adfoc.us left intact
```

Now, using the ubuntu container which gives the expected behavior:
```
* TCP_NODELAY set
* Connected to adfoc.us (10.9.8.181) port 80 (#0)
> GET /google.com HTTP/1.1
> Host: adfoc.us
> User-Agent: curl/7.56.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< X-Pi-hole: A black hole for Internet advertisements.
< Content-type: text/html; charset=UTF-8
< Transfer-Encoding: chunked
< Date: Sun, 29 Oct 2017 14:57:44 GMT
< Server: lighttpd/1.4.35
< 
{ [179 bytes data]
100   168    0   168    0     0    168      0 --:--:-- --:--:-- --:--:-- 42000
* Connection #0 to host adfoc.us left intact
<html>
<head>
<script>window.close();</script></head>
<body>
<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7">
</body>
</html>
```

**Comment by graysky2** [comment]
Nevermind... the problem was with the nginx.conf and the need to edit admin/index.php as detailed here: https://github.com/pi-hole/pi-hole/wiki/Nginx-Configuration

**Comment by jacobsalmela** [comment]
Resolved in v3.2 (#1602)
